docs/adrs/004-SplitAndAggregate.md

```diff
@@ -0,0 +1,139 @@
+# Replace KBParallels with another solution to avoid Deadlocks
+
+Date: 2021-09-22
+
+[Related ADR](https://github.com/kbase/execution_engine2/blob/develop/docs/adrs/rework-batch-analysis-architecture.md)
+
+## Note
+This ADR is more of a place to keep the current discussions we had at https://docs.google.com/document/d/1AWjayMoqCoGkpO9-tjXxEvO40yYnFtcECbdne5vTURo
+Rather than to make a decision. There is still more planning, scoping and testing involved before we can fully design this system.
+
+Still to be determined (not in scope of this ADR):
+* UI and how it relates to the bulk execution
+* XSV Analysis and how it relates to the bulk execution
+
+
+## Intro
+Sometimes a calculation requires too many resources from one node (walltime, memory, disk), so the calculation gets spread across multiple machines.
+The final step of the app that uses KBParallels is to create a report. This step may use results from all of the computed jobs to create the final report.
+In order to do this, some apps use a mechanism called KBParallel. The apps are listed at [DATAUP-111](https://kbase-jira.atlassian.net/browse/DATAUP-111)
+
+## The current implementation of Batch Analysis in [kb_BatchApp](https://github.com/kbaseapps/kb_BatchApp) at KBase has the following issues:
+
+* Current UI is not adequate: Users shouldn’t have to code in order to run batch analysis. Also it’s difficult to do so, even for those familiar with KBase code (have to find object names)
+* Dependency on [KBParallel](https://github.com/kbaseapps/KBParallel): any changes to KBParallel could affect KB Batch and subsequently all other apps. 
+* Queue deadlocking: users have a max of 10 slots in the queue, with the current implementation one is taken up just to manage the jobs. Could lead to deadlock scenarios
```

Comment:
Can you quickly explain how this happens (e.g. 10 'management' jobs start before all the jobs for at least one of the management jobs finish, and so there are 10 management jobs running, and no 'real' jobs can start. Since the management jobs can't finish until their real jobs finish, the queue is deadlocked).

Reply:
done

Reply:
:+1: